File: cpp/src/domain/data_generation/mandelbrot_generator.cpp

#include <cmath>
#include "mandelbrot_generator.h"
#include <iostream>


ArrayXi MandelbrotGenerator::computeMandelbrot(ArrayXcf const & c, unsigned short const & max, unsigned short const & threshold) {
    ArrayXi result = ArrayXi::Zero(c.size());
    ArrayXcf z(c.size());
    z << c;
    unsigned short i = 1;
    Array<bool, Dynamic, 1> mask(c.size());

    while (i < max) {
        mask = result.cwiseEqual(0);
        if (mask.sum() == 0) {
            break;
        }
        z = z.square() + c;

        mask = mask && ((z.real() * z.imag()) > threshold);
        result += mask.cast<int>() * i;
        i++;
    }
    result += (1 - result.cast<bool>().cast<int>()) * max;
    return result;
}

ArrayXXf MandelbrotGenerator::get(int nRows) {
    int rootN = static_cast<int>(ceil(sqrt(nRows)));
    ArrayXf real = ArrayXf::LinSpaced(rootN, R_MIN, R_MAX)
            .replicate(rootN, 1);
    ArrayXf imag = ArrayXf::LinSpaced(rootN, I_MIN, I_MAX)
            .replicate(rootN, 1);
    std::sort(real.begin(), real.end(), std::less<>());

    long size = real.size();
    ArrayXcf data = ArrayXcf::Zero(size, 1);
    data += real;
    data.imag() += imag;
    ArrayXi const y = computeMandelbrot(data, max, threshold);

    ArrayXXf result(size, 3);
    result.col(0) << data.real();
    result.col(1) << data.imag();
    result.col(2) << y.cast<float>();
    return result;
}

File: python/mlops_examples/domain/data_generation/mandelbrot_generator.py
import numpy as np
import pandas as pd

from mlops_examples.domain.data_generation.data_generator import DataGenerator


class MandelbrotGenerator(DataGenerator):
    R_MIN = -2
    R_MAX = 0.47
    I_MIN = -1.12
    I_MAX = 1.12

    def __init__(self):
        self.__max: int = 25
        self._threshold: int = 4

    def with_max(self, m: int) -> 'MandelbrotGenerator':
        self.__max = m
        return self

    def get(self, n_rows: int) -> pd.DataFrame:
        data = np.empty(n_rows, dtype=np.complex64)
        root_n = np.ceil(np.sqrt(n_rows)).astype(int)
        data[:].real = np.repeat(np.linspace(self.R_MIN, self.R_MAX, root_n), root_n)[:n_rows]
        data[:].imag = np.tile(np.linspace(self.I_MIN, self.I_MAX, root_n), root_n)[:n_rows]
        result = self.__compute_mandelbrot(data)
        df = pd.DataFrame(
            columns=["r", "i", "target"],
            data=np.stack((data.real, data.imag, result), axis=1)
        )
        return df

    def __compute_mandelbrot(self, c: np.ndarray) -> np.ndarray:
        result = np.zeros_like(c, dtype=int)
        z = np.copy(c)
        i = 1
        while i < self.__max:
            unset = result == 0
            if unset.sum() < 1:
                break
            z = np.square(z) + c
            mask = unset & (z.real * z.imag > self._threshold)
            result[mask] = i
            i += 1
        unset = result == 0
        result[unset] = self.__max
        return result
